luarules/gadgets/game_initial_spawn.lua

```diff
@@ -114,6 +118,220 @@ if gadgetHandler:IsSyncedCode() then
+				orderMsg = tname
+				orderIds = FindPlayerIDFromTeamID(teamID)
+			else
+				if alone then alone = false end
+				orderMsg = orderMsg .. ", " .. tname
+				orderIds = orderIds .. " " .. FindPlayerIDFromTeamID(teamID)
+			end
+		end
+		if log and not alone then
+			Spring.Log(gadget:GetInfo().name, LOG.INFO, "Order [id:"..allyTeamID_ready.."]: "..orderMsg)
+		end
+		Spring.SendLuaUIMsg("DraftOrderPlayersOrder " .. allyTeamID_ready .. " " .. orderIds)
+	end
+
+	local function calculateVotedPercentage(allyTeamID)
+		local totalPlayers = Spring.GetTeamCount(allyTeamID)
+		local votedPlayers = 0
+		for teamID, _ in pairs(votedToForceSkipTurn[allyTeamID]) do
+			if votedToForceSkipTurn[allyTeamID][teamID] then
+				votedPlayers = votedPlayers + 1
+			end
+		end
+		return (votedPlayers / totalPlayers) * 100
+	end
+
+	local function checkVotesAndAdvanceTurn(allyTeamID)
+		if allyTeamSpawnOrderPlaced[allyTeamID] >= #allyTeamSpawnOrder[allyTeamID] then return end
+		local votedPercentage = calculateVotedPercentage(allyTeamID)
+		if votedPercentage > 50 then
+			--Spring.Echo("Over 50% of players on ally team " .. allyTeamID .. " have voted.")
+			for teamID, _ in pairs(votedToForceSkipTurn[allyTeamID]) do
+				votedToForceSkipTurn[allyTeamID][teamID] = false -- reset vote
+			end -- and advance turn
+			allyTeamSpawnOrderPlaced[allyTeamID] = allyTeamSpawnOrderPlaced[allyTeamID]+1
+			SendDraftMessageToPlayer(allyTeamID, allyTeamSpawnOrderPlaced[allyTeamID])
+		end
+	end
+
+	local function isTurnToPlace(allyTeamID, teamID)
+		-- Returns:
+		-- 0 - Can't place yet
+		-- 1 - Your turn
+		-- 2 - Your turn has passed
+		local teamOrder = allyTeamSpawnOrder[allyTeamID]
+		if not teamOrder then
+			return 2  -- No spawn order defined for this team; you can place whenever then
+		end
+		local placedIndex = allyTeamSpawnOrderPlaced[allyTeamID]
+		if placedIndex <= 0 then
+			return 2 -- Cannot figure out whose turn; you can place then
+		end
+		local teamIndex = nil
+		-- Find the index of the team within the spawn order
+		for i, id in ipairs(teamOrder) do
+			if id == teamID then
+				teamIndex = i
+				break
+			end
+		end
+		if not teamIndex then return 2 end -- The team is not in the spawn order; you can place whenever then
+		if teamIndex == placedIndex then -- Your turn
+			return 1
+		elseif teamIndex < placedIndex then -- Skipped your turn?
+			return 2
+		else
+			return 0 -- Your turn is yet to be
+		end
+	end
+
+	local function SendDraftMessageToPlayer(allyTeamID, target_num)
+		-- we send allyTeamID currentTurn nextTurn index
+		if target_num + 1 <= #allyTeamSpawnOrder[allyTeamID] then
+			local playerID_draft = FindPlayerIDFromTeamID(allyTeamSpawnOrder[allyTeamID][target_num])
+			local playerID_next_in_queue_draft = FindPlayerIDFromTeamID(allyTeamSpawnOrder[allyTeamID][target_num + 1])
+			Spring.SendLuaUIMsg("DraftOrderPlayerTurn " .. allyTeamID .. " " .. playerID_draft .. " " .. playerID_next_in_queue_draft .. " " .. target_num)
```

Comment:
I would like to see these lua messages formalized somewhat, there is a lot of potential for typo errors here, with the precise formatting and key strings it requires.

Reply:
Yeah this was first coded even before we started sending order index. So I've removed 2 params out of 4.